Moving text on spinbit into transport (#2364)
quicwg/base-drafts: draft-ietf-quic-transport.md

    target:
     "https://web.archive.org/web/20150315054838/http://ha.ckers.org/slowloris/"

  QUIC-MANAGEABILITY:
    title: "Manageability of the QUIC Transport Protocol"
    author:

>       - ins: M. Kuehlewind

Reviewer: add `-ins: B. Trammell` as well

of a frame type that uses a longer encoding than necessary as a connection error
of type PROTOCOL_VIOLATION.

## Latency Spin Bit {#spin-bit}

The latency spin bit enables passive latency monitoring from observation points
on the network path throughout the duration of a connection. The spin bit is
only present in the short packet header, since it is possible to measure the
initial RTT of a connection by observing the handshake. Therefore, the spin bit

> will appear after version negotiation and connection establishment are

Reviewer 2: will appear -> is available

completed. On-path measurement and use of the Latency Spin Bit is further
discussed in {{QUIC-MANAGEABILITY}}.

> The spin bit utilizes a single bit in the first byte of the short header. The

Reviewer: +1

location of the bit and procedures for how to set it by clients and servers are
defined in {{short-header}}.

> Implementations MAY select to not implement the full spin bit functionality. In

Reviewer: I think I'd phrase this differently: "the spin bit is an optional feature. implementations that implement MUST implement it as specified in this document."

Author: Agree, that is a better phrasing

Reviewer 2: Agree, too. Would use the 2119 OPTIONAL capitalization

that case they are only REQUIRED to implement what is defined for the spin bit
when it is disabled.

Each endpoint unilaterally decides if the spin bit is enabled or disabled for a
connection. Implementations SHOULD allow administrators of clients and servers
to disable the spin bit either globally or on a per-connection basis. Even when

> the spin bit is not disabled by the administrator implementations SHOULD disable

Reviewer: IIRC consensus here sounded like MUST probabalistically disable, since we will want to provide cover traffic in the long term, not only for non-implementing endpoints but also for flows that decide to disable it. I'd justify the MUST because it affects the privacy properties of the wire image (however minimal the actual effect is)

```suggestion
the spin bit is not disabled by the administrator implementations MUST disable
```

Author: >           Also, we need to start an issue on this rather than commenting on the PR
@larseggert I'm not sure what the issue is? 
The consensus in BKK was that the bit should be included in the standard, this PR only contains the information necessary for implementors, no text on how to interpret or process the bit on path. Simply referencing a non-milestone potentially experimental document seems a bit odd.

the spin bit on a randomly chosen fraction of connections. However, connections

> may be configured to explicitly enable spinning, for example in the case of

Reviewer: in line with the SHOULD above, this isn't a may/MAY:

```suggestion
SHOULD be configurable to explicitly enable spinning, for example in the case of
```

Author: Not sure why this would be a SHOULD, here we describe a case where a single connection has been explicitly configured to spin (e.g some debug or tracing mode enabled). In that case, spinning shouldn't be disabled for that specific connection as a result of the random selection process.

> explicit customer support and debugging.

Reviewer 2: I'm not sure we had consensus for that, because if an app (erroneously) enables spin for all connections using this API, there would be no cover traffic, and we required there to always be some non-spin traffic.

Spin Bit (S):

> : The third most significant bit (0x20) of byte 0 is the Latency Spin Bit.

Reviewer 3: I think this would be easier to read if either all the text was here or if this referenced the section above.  Given it's a lot of text, I tend to think the latter is better.

Reviewer: +1 to this

> : When the spin bit is disabled, endpoints MAY set the spin bit to any value,

Reviewer 4: Please don't have so many paragraphs nested like this.  A forward reference to a section that includes this text would be much better.

and MUST accept any incoming value. It is RECOMMENDED that they
set the spin bit to a random value either chosen independently for each packet,
or chosen independently for each path and kept constant for that path.

> : If the spin bit is enabled for the connection the endpoint maintains a spin

Reviewer 4: ```suggestion
: If the spin bit is enabled for the connection, the endpoint maintains a spin
```

> value, 0 or 1, and sets the spin bit in the short header to the currently stored

Reviewer 4: ```suggestion
value and sets the spin bit in the short header to the currently stored
```

value when a packet with a short header is sent out. The spin value is
initialized to 0 in the endpoint at connection start.  Each endpoint also
remembers the highest packet number seen from its peer on the connection.

> : When a server receives a packet from a client, if that packet has a short

Reviewer 2: Simply say "when a server receives a short-header packet"?

header and if it increments the highest packet number seen by the server from
the client, the server sets the spin value to the value observed in the spin bit
in the received packet.

> : When a client receives a packet from a server, if the packet has a short

Reviewer 2: Ditto

versions of QUIC are interpreted.

> 

Reviewer 4: ```suggestion
```
